pkg/controller/factory/base_controller.go

```diff
@@ -66,6 +68,24 @@ func (c *baseController) Run(ctx context.Context, workers int) {
 		}()
 	}
 
+	// run post-run hooks (custom triggers, etc.)
+	if len(c.postRunHooks) > 0 {
+		var hookWaitGroup sync.WaitGroup
+		defer func() {
+			defer klog.Infof("All %s post-run hooks have been terminated", c.name)
+			hookWaitGroup.Wait() // wait for the post-run hooks
+		}()
+		for i := range c.postRunHooks {
+			hookWaitGroup.Add(1)
+			go func() {
+				defer hookWaitGroup.Done()
+				if err := c.postRunHooks[i](ctx, c.syncContext); err != nil {
```

Comment:
`i` race. Make it a param of the go func.

Reply:
good catch, I will add some unit-test as follow up.

Reply:
fixed.